File: ios/RNAmazonTAMInterstitial.m
#import <Foundation/Foundation.h>
#import "RNAmazonTAMInterstitial.h"

@interface RNAmazonTAMInterstitial ()

@property (nonatomic, strong) RCTPromiseResolveBlock resolve;
@property (nonatomic, strong) RCTPromiseRejectBlock reject;

@end

@implementation RNAmazonTAMInterstitial

RCT_EXPORT_MODULE(AmazonTAMInterstitial);

- (dispatch_queue_t)methodQueue
{
  return dispatch_get_main_queue();
}

#pragma mark - Interstitial Ad
RCT_REMAP_METHOD(requestInterstitialAd,
                 slotUUID:(NSString *)slotUUID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  _resolve = resolve;
  _reject = reject;
  DTBAdSize *size = [[DTBAdSize alloc] initInterstitialAdSizeWithSlotUUID:slotUUID];
  DTBAdLoader *adLoader = [DTBAdLoader new];
  [adLoader setSizes:size, nil];
  [adLoader loadAd:self];
}

#pragma mark - DTBAdCallback
- (void)onFailure: (DTBAdError)error {
  NSString *message = @"Failed to load interstitial ad from Amazon.";
  NSError *errorResponse = [[NSError alloc] initWithDomain:@"RNAmazonTAMInterstitial" code:error userInfo:@{NSLocalizedDescriptionKey: message}];
  if (_reject) {
    _reject(@"FAILED_TO_LOAD", message, errorResponse );
    _reject = nil;
  }
}

- (void)onSuccess: (DTBAdResponse *)adResponse {
  if (_resolve) {
    _resolve(adResponse.customTargeting);
    _resolve = nil;
  }
}

@end

File: android/src/main/java/com/krazylabs/rnamazontam/RNAmazonTAMInterstitial.java
package com.krazylabs.rnamazontam;

import android.util.Log;

import com.amazon.device.ads.AdError;
import com.amazon.device.ads.DTBAdCallback;
import com.amazon.device.ads.DTBAdRequest;
import com.amazon.device.ads.DTBAdResponse;
import com.amazon.device.ads.DTBAdSize;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;

public class RNAmazonTAMInterstitial extends ReactContextBaseJavaModule {
  private final ReactApplicationContext reactContext;

  public RNAmazonTAMInterstitial(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  @Override
  public String getName() {
    return "AmazonTAMInterstitial";
  }

  @ReactMethod
  public void requestInterstitialAd(String slotUUID, final Promise promise) {
    final DTBAdRequest adLoader = new DTBAdRequest();
    adLoader.setSizes(new DTBAdSize.DTBInterstitialAdSize(slotUUID));
    adLoader.loadAd(new DTBAdCallback() {
      @Override
      public void onFailure(AdError adError) {
        String code = adError.getCode().toString();
        String message = "Failed to load interstitial ad from Amazon: " + adError.getMessage();
        Log.e(code, message);
        promise.reject(code, message);
      }

      @Override
      public void onSuccess(DTBAdResponse dtbAdResponse) {
        WritableMap customTargeting = RNAmazonTAMUtils.getCustomTargeting(dtbAdResponse);
        promise.resolve(customTargeting);
      }
    });
  }
}
